File: app/src/main/java/com/example/savaari/Repository.java
package com.example.savaari;

import android.util.Log;

import com.example.savaari.services.network.NetworkUtil;
import com.example.savaari.services.network.OnDataLoadedListener;

import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.Properties;
import java.util.concurrent.Executor;

public class Repository {
    private static String url;
    private Executor executor;

    Repository(Executor executor) {
        this.executor = executor;
        url = loadDataSourceUrl();
        Log.d("Repository url: ", url);
    }

    public String loadDataSourceUrl() {
        Properties prop;
        String propFileName = "config.properties";
        InputStream inputStream;

        try {
            prop = new Properties();
            inputStream = getClass().getClassLoader().getResourceAsStream(propFileName);

            if (inputStream != null) {
                prop.load(inputStream);
            } else {
                throw new FileNotFoundException("property file '" + propFileName + "' not found in the classpath");
            }

            // Get property value
            return prop.getProperty("dataSourceUrl");
        }
        catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }


    //Find Driver
    public void findDriver(OnDataLoadedListener callback, int currentUserID, double srcLatitude,
                           double srcLongitude, double destLatitude, double destLongitude, int paymentMode, int rideType) {
        
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().findDriver(url, currentUserID,
                srcLatitude, srcLongitude, destLatitude, destLongitude, paymentMode, rideType)));
    }

    // Sign-Up
    public void signup(OnDataLoadedListener callback, String nickname, String username, String password) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().signup(url, nickname, username, password)));
    }
    // Login
    public void login(OnDataLoadedListener callback, String username, String password) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().login(url, username, password)));
    }

    public void persistLogin(OnDataLoadedListener callback, Integer userID) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().persistLogin(url, userID)));
    }

    public void logout(OnDataLoadedListener callback, Integer userID) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().logout(url, userID)));
    }
    // Loading User Data
    public void loadUserData(OnDataLoadedListener callback, int currentUserID) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().loadUserData(url, currentUserID)));
    }

    // Get User Locations
    public void getUserLocations(OnDataLoadedListener callback) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().getUserLocations(url)));
    }

    // Send Last Location
    public void sendLastLocation(int currentUserID, double latitude, double longitude) {
        executor.execute(() ->
                NetworkUtil.getInstance().sendLastLocation(url, currentUserID, latitude, longitude));
    }

    public void getDriverLocation(OnDataLoadedListener callback, int driverID) {
        callback.onDataLoaded(NetworkUtil.getInstance().getDriverLocation(url, driverID));
    }

    public void getRide(OnDataLoadedListener callback, int riderID) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().getRide(url, riderID)));
    }

    public void getRideStatus(OnDataLoadedListener callback, int rideID) {
        callback.onDataLoaded(NetworkUtil.getInstance().getRideStatus(url, rideID));
    }

    public void acknowledgeEndOfRide(OnDataLoadedListener callback, int rideID, int riderID) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().acknowledgeEndOfRide(url, rideID, riderID)));
    }

    public void giveFeedbackForDriver(OnDataLoadedListener callback, int rideID, int driverID, float rating) {
        executor.execute(() -> callback.onDataLoaded(NetworkUtil.getInstance().giveFeedbackForDriver(url, rideID, driverID, rating)));
    }
}

File: app/src/main/java/com/farjad/savaari/services/persistence/Repository.kt
package com.farjad.savaari.services.persistence

import android.util.Log
import androidx.lifecycle.LiveData
import com.farjad.savaari.services.network.OnDataLoadedListener
import com.farjad.savaari.services.network.NetworkUtil
import com.farjad.savaari.services.network.WebService
import com.farjad.savaari.services.persistence.user.UserDao
import com.farjad.savaari.services.persistence.user.UserDataModel
import java.io.FileNotFoundException
import java.io.InputStream
import java.lang.Exception
import java.util.*
import java.util.concurrent.Executor
import java.util.concurrent.TimeUnit

class Repository internal constructor(private val executor: Executor,
                                      private val webService: WebService,
                                      private val userDao: UserDao
                                      )
{

    private var url : String = "";

    init {
        url = loadDataSourceUrl()
        Log.d("Repository url: ", url)
    }

    companion object {
        val FRESH_TIMEOUT = TimeUnit.DAYS.toMillis(1)
        val TAG = Repository::class.java.name
    }

    private fun loadDataSourceUrl(): String {
        val prop: Properties
        val propFileName = "config.properties"
        val inputStream: InputStream?
        return try {
            prop = Properties()
            inputStream = javaClass.classLoader!!.getResourceAsStream(propFileName)
            if (inputStream != null) {
                prop.load(inputStream)
            } else {
                throw FileNotFoundException("property file '$propFileName' not found in the classpath")
            }

            // Get property value
            prop.getProperty("dataSourceUrl")
        } catch (e: Exception) {
            e.printStackTrace()
            ""
        }
    }


    fun getUser(userId: Int): LiveData<UserDataModel> {
        refreshUser(userId)
        // Returns a LiveData object directly from the database.
        return userDao.load(userId)
    }

    private fun refreshUser(userID : Int) {

        Log.d(TAG, "refreshUser called!");
        // Runs in a background thread.
        executor.execute {
            // Check if user data was fetched recently.
            val userExists = userDao.hasUser(userID, FRESH_TIMEOUT)

            if (userExists > 0) {
                Log.d(TAG, "loadRoomData: User does not exist, refreshing data")
                // Refreshes the data.
                val response = webService.getUser(userID).execute()

                // Check for errors here.

                // Updates the database. The LiveData object automatically
                // refreshes, so we don't need to do anything else here.
                userDao.save(response.body()!!)
            }
            else {
                Log.d(TAG, "loadRoomData: User already exists!")
            }
        }
    }

    //Find Driver
    fun findDriver(callback: OnDataLoadedListener, currentUserID: Int, srcLatitude: Double,
                   srcLongitude: Double, destLatitude: Double, destLongitude: Double, paymentMode: Int, rideType: Int) {
        executor.execute {
            callback.onDataLoaded(NetworkUtil.getInstance().findDriver(url, currentUserID,
                    srcLatitude, srcLongitude, destLatitude, destLongitude, paymentMode, rideType))
        }
    }

    // Sign-Up
    fun signup(callback: OnDataLoadedListener, nickname: String?, username: String?, password: String?) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().signup(url, nickname, username, password)) }
    }

    // Login
    fun login(callback: OnDataLoadedListener, username: String?, password: String?) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().login(url, username, password)) }
    }

    fun persistLogin(callback: OnDataLoadedListener, userID: Int?) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().persistLogin(url, userID)) }
    }

    fun logout(callback: OnDataLoadedListener, userID: Int?) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().logout(url, userID)) }
    }

    // Loading User Data
    fun loadUserData(callback: OnDataLoadedListener, currentUserID: Int) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().loadUserData(url, currentUserID)) }
    }

    // Get User Locations
    fun getUserLocations(callback: OnDataLoadedListener) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().getUserLocations(url)) }
    }

    // Send Last Location
    fun sendLastLocation(currentUserID: Int, latitude: Double, longitude: Double) {
        executor.execute { NetworkUtil.getInstance().sendLastLocation(url, currentUserID, latitude, longitude) }
    }

    fun getDriverLocation(callback: OnDataLoadedListener, driverID: Int) {
        callback.onDataLoaded(NetworkUtil.getInstance().getDriverLocation(url, driverID))
    }

    fun getRide(callback: OnDataLoadedListener, riderID: Int) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().getRide(url, riderID)) }
    }

    fun getRideStatus(callback: OnDataLoadedListener, rideID: Int) {
        callback.onDataLoaded(NetworkUtil.getInstance().getRideStatus(url, rideID))
    }

    fun acknowledgeEndOfRide(callback: OnDataLoadedListener, rideID: Int, riderID: Int) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().acknowledgeEndOfRide(url, rideID, riderID)) }
    }

    fun giveFeedbackForDriver(callback: OnDataLoadedListener, rideID: Int, driverID: Int, rating: Float) {
        executor.execute { callback.onDataLoaded(NetworkUtil.getInstance().giveFeedbackForDriver(url, rideID, driverID, rating)) }
    }
}